provider/maas/environ.go

```diff
@@ -948,6 +951,33 @@ func (environ *maasEnviron) StartInstance(args environs.StartInstanceParams) (
 			return nil, errors.Annotatef(err, "cannot run instances")
 		}
 	}
+	done := make(chan struct{})
+	defer close(done)
+	go func(done chan struct{}) {
```

Comment:
Some points here. I agree entirely that this method is too long and should be broken up. Please don't add big bunches of code to the middle.

Don't even start the polling goroutine if there is no StatusCallback. Otherwise we are going to see repeated log entries for something.

Don't hide time.After behind clock.WallClock. Either use the clock interface in the environ or don't, but don't pretend.

Since this is a continuous poll rather than a retry operation, I'm happy for it to be in a for loop that is exited when done, but please define it in a function outside this one.

Don't take the done channel as `chan struct{}` but the more restrictive `<-chan struct{}`. You are never sending to it.